paddle/fluid/platform/enforce.h

```diff
@@ -366,28 +367,72 @@ using CommonType1 = typename std::add_lvalue_reference<
 template <typename T1, typename T2>
 using CommonType2 = typename std::add_lvalue_reference<
     typename std::add_const<typename TypeConverter<T1, T2>::Type2>::type>::type;
+
+// Here, we use SFINAE to check whether T can be converted to std::string
+template <typename T>
+class CanToString {
+ private:
+  using YesType = uint8_t;
+  using NoType = uint16_t;
+
+  template <typename U>
+  static YesType Check(decltype(std::cout << std::declval<U>())) {
+    return 0;
+  }
+
+  template <typename U>
+  static NoType Check(...) {
+    return 0;
+  }
+
+ public:
+  static constexpr bool kValue =
+      std::is_same<YesType, decltype(Check<T>(std::cout))>::value;
+};
+
+template <bool kCanToString /* = true */>
```

Comment:
Two questions:

1. Can we change "struct" to "class"? Use a struct only for passive objects that carry data; everything else is a class. Ref: 
https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes 

2. After we get "constexpr bool __kCanToString__", are we able to use "if" clause without template class and function?

Reply:
1. Done.

2. We can only use template classes. Consider the following codes using "if" clause:
```C++
std::string Convert(const char *expression, const T& value, bool can_to_string) {
    if (can_to_string) {
       return expression + std::string(":") + string::to_string(value); // Compilation error if T cannot be converted to std::string!
    } else {
       return expression;
    } 
}
```
The codes above would cause compilation error!

Reply:
Discussed offline, I meant something like #if macro. Sorry for the confusion. We found a case that we can use #if, but we are afraid of Windows supports different C++ standard ... so we won't change here.